python/python/lance/progress.py

```diff
@@ -0,0 +1,162 @@
+#  Copyright (c) 2023. Lance Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+# ruff: noqa: F821
+
+from __future__ import annotations
+
+import json
+from abc import ABC, abstractmethod
+from typing import Dict, Optional
+
+
+class FragmentWriteProgress(ABC):
+    """Progress tracking for Writing a Dataset or Fragment.
+
+    Warns
+    -----
+    This tracking class is experimental and may change in the future.
+    """
+
+    def _do_begin(
+        self, fragment_json: str, multipart_id: Optional[str] = None, **kwargs
+    ):
+        """Called when a new fragment is created"""
+        from .fragment import FragmentMetadata
+
+        fragment = FragmentMetadata.from_json(fragment_json)
+        return self.begin(fragment, multipart_id, **kwargs)
+
+    @abstractmethod
+    def begin(
+        self, fragment: "FragmentMetadata", multipart_id: Optional[str] = None, **kwargs
+    ) -> None:
+        """Called when a new fragment is created.
+
+        Parameters
+        ----------
+        fragment : FragmentMetadata
+            The fragment that is open to write to.
+        multipart_id : str, optional
+            The multipart id to upload this fragment to cloud storage.
```

Comment:
```suggestion
            The multipart id that will be uploaded to cloud storage. This may be
            used later to abort incomplete uploads if this fragment write fails.
```